client/src/main/scala/org/apache/celeborn/client/ShuffleClientImpl.scala

```diff
@@ -0,0 +1,1121 @@
+
+  private val mapperEndMap: ConcurrentHashMap[Integer, Set[String]] =
+    new ConcurrentHashMap[Integer, Set[String]]
+
+  // key: shuffleId-mapId-attemptId
+  private val pushStates: Map[String, PushState] = new ConcurrentHashMap[String, PushState]
+
+  private var pushDataRetryPool: ExecutorService = null
+
+  private var partitionSplitPool: ExecutorService = null
+  private val splitting: Map[Integer, Set[Integer]] = new ConcurrentHashMap[Integer, Set[Integer]]
+
+  private[client] val compressorThreadLocal: ThreadLocal[Compressor] =
+    new ThreadLocal[Compressor]() {
+      override protected def initialValue: Compressor = {
+        Compressor.getCompressor(conf)
+      }
+    }
+
+  // key: shuffleId
+  private val reduceFileGroupsMap: Map[Integer, ReduceFileGroups] =
+    new ConcurrentHashMap[Integer, ReduceFileGroups]
+
+  def this(conf: RssConf, userIdentifier: UserIdentifier) {
+    this()
+    this.conf = conf
+    this.userIdentifier = userIdentifier
+    registerShuffleMaxRetries = RssConf.registerShuffleMaxRetry(conf)
+    registerShuffleRetryWait = RssConf.registerShuffleRetryWait(conf)
+    maxInFlight = RssConf.pushDataMaxReqsInFlight(conf)
+    pushBufferSize = RssConf.pushDataBufferSize(conf)
+    // init rpc env and master endpointRef
+    rpcEnv = RpcEnv.create("ShuffleClient", Utils.localHostName, 0, conf)
+    val dataTransportConf: TransportConf = Utils.fromRssConf(
+      conf,
+      TransportModuleConstants.DATA_MODULE,
+      conf.getInt("rss.data.io.threads", 8))
+    val context: TransportContext =
+      new TransportContext(dataTransportConf, new BaseMessageHandler, true)
+    dataClientFactory = context.createClientFactory
+    val retryThreadNum: Int = RssConf.pushDataRetryThreadNum(conf)
+    pushDataRetryPool = ThreadUtils.newDaemonCachedThreadPool("Retry-Sender", retryThreadNum, 60)
+    val splitPoolSize: Int = RssConf.clientSplitPoolSize(conf)
+    partitionSplitPool = ThreadUtils.newDaemonCachedThreadPool("Shuffle-Split", splitPoolSize, 60)
+  }
+
+  private def submitRetryPushData(
+      applicationId: String,
+      shuffleId: Int,
+      mapId: Int,
+      attemptId: Int,
+      body: Array[Byte],
+      batchId: Int,
+      loc: PartitionLocation,
+      callback: RpcResponseCallback,
+      pushState: PushState,
+      cause: StatusCode): Unit = {
+    val partitionId: Int = loc.getId
+    if (!revive(
+        applicationId,
+        shuffleId,
+        mapId,
+        attemptId,
+        partitionId,
+        loc.getEpoch,
+        loc,
+        cause)) {
+      callback.onFailure(new IOException("Revive Failed"))
+    } else {
+      if (mapperEnded(shuffleId, mapId, attemptId)) {
+        logDebug(s"Retrying push data, but the mapper(map $mapId attempt $attemptId) has ended.")
+        pushState.inFlightBatches.remove(batchId)
+      } else {
+        val newLoc: PartitionLocation = reducePartitionMap.get(shuffleId).get(partitionId)
```

Comment:
I think the type can be removed to be more concise, multiple places.

Reply:
> I think the type can be removed to be more concise, multiple places.

Sure